### ISSUE TITLE
BPF host routing bypasses DNS resolution of dockerd on Kind clusters

### ISSUE THREAD

**Issue by brb** [cilium/cilium#23283]
While debugging https://github.com/cilium/cilium/pull/23171, I've discovered that the BPF masquerade is not properly working on > 5.4 kernels when running in the `ci-datapath`. The main symptom was failing requests to `https://one.one.one.one`. A closer look revealed that the DNS resolution for external FQDN was failing.

With the help of tcpdump running on the node of the CoreDNS pods (`kind-control-plane`), with the iptables masquerading I got the following trace:

```
eth0  In  IP 10.244.1.217.39869 > 10.244.0.74.53: 33504+ [1au] A? googlel.com. (52)
lxc61ce9e2c0db2 Out IP 10.244.1.217.39869 > 10.244.0.74.53: 33504+ [1au] A? googlel.com. (52)
lxc61ce9e2c0db2 In  IP 10.244.0.74.40127 > 172.12.1.1.53: 57830+ [1au] A? googlel.com. (52)
eth0  Out IP 172.12.1.4.40235 > 8.8.8.8.53: 57830+ [1au] A? googlel.com. (52)
eth0  In  IP 8.8.8.8.53 > 172.12.1.4.40235: 57830 0/1/1 (122)
```

While with the BPF masquerading - the following:
```
eth0  In  IP 10.244.1.217.42393 > 10.244.0.74.53: 6280+ [1au] A? googlel.com. (52)
lxc61ce9e2c0db2 In  IP 10.244.0.74.57566 > 172.12.1.1.53: 46622+ [1au] A? googlel.com. (52)
eth0  Out IP 172.12.1.4.57566 > 172.12.1.1.53: 46622+ [1au] A? googlel.com. (52)
```

- `10.244.1.217` - the client pod.
- `10.244.0.74` - the CoreDNS pod (`lxc61ce9e2c0db2`).
- `172.12.1.4`  - the CoreDNS node IP (`eth0`).
- `172.12.1.1` - the Docker bridge IP addr which connects Kind nodes.

From the traces diff it quickly became clear that in the end that it's not the BPF masq matter, but the BPF host routing (with the iptables masquerading it's automatically disabled), and that the following Docker's DNS xlation is bypassed with the BPF host routing:
```
lxc61ce9e2c0db2 In  IP 10.244.0.74.40127 > 172.12.1.1.53: 57830+ [1au] A? googlel.com. (52)
eth0  Out IP 172.12.1.4.40235 > 8.8.8.8.53: 57830+ [1au] A? googlel.com. (52)
```

I still don't have a clue how the translation above happens.

**Comment by brb** [comment]
Looking at the filtered strace of `dockerd` running in the host netns (:exploding_head:):
```
1154  setsockopt(28<UDP:[3318259]>, SOL_SOCKET, SO_BROADCAST, [1], 4 <unfinished ...>
1154  <... setsockopt resumed>)         = 0
1154  connect(28<UDP:[3318259]>, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, 16 <unfinished .
1154  <... connect resumed>)            = 0
1154  epoll_ctl(5<anon_inode:[eventpoll]>, EPOLL_CTL_ADD, 28<UDP:[3318259]>, {events=EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, data={u32=3036353360, u64=139898711118672}} <unfinished ...>
1081  <... nanosleep resumed>NULL)      = 0
1154  <... epoll_ctl resumed>)          = 0
1106  <... epoll_pwait resumed>[{events=EPOLLOUT, data={u32=3036353360, u64=139898711118672}}], 128, 8, NULL, 63740320030353) = 1
1081  nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
1154  getsockname(28<UDP:[3318259]>,  <unfinished ...>
1106  epoll_pwait(5<anon_inode:[eventpoll]>,  <unfinished ...>
1081  <... nanosleep resumed>NULL)      = 0
1106  <... epoll_pwait resumed>[], 128, 0, NULL, 0) = 0
1154  <... getsockname resumed>{sa_family=AF_INET, sin_port=htons(37754), sin_addr=inet_addr("172.12.1.4")}, [112 => 16]) = 0
1081  nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
1106  epoll_pwait(5<anon_inode:[eventpoll]>,  <unfinished ...>
1154  getpeername(28<UDP:[3318259]>,  <unfinished ...>
1081  <... nanosleep resumed>NULL)      = 0
1081  nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
1154  <... getpeername resumed>{sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, [112 => 16]) = 0
1081  <... nanosleep resumed>NULL)      = 0
1154  setns(14<net:[4026531840]>, CLONE_NEWNET <unfinished ...>
1081  nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
1154  <... setns resumed>)              = 0
1081  <... nanosleep resumed>NULL)      = 0
1154  close(26</run/docker/netns/acf7be30eece> <unfinished ...>
1081  nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
1154  <... close resumed>)              = 0
1081  <... nanosleep resumed>NULL)      = 0
1154  write(28<UDP:[3318259]>, "\261\252\1\0\0\1\0\0\0\0\0\1\7googlel\3com\0\0\34\0\1\0\0)"..., 40 <unfinished ...>
```

**Comment by brb** [comment]
Looking into the filtered strace output of one of the `dockerd` threads which handles requests:
```
1154  recvmsg(69<UDP:[691492]>,  <unfinished ...>
1154  <... recvmsg resumed>{msg_name={sa_family=AF_INET, sin_port=htons(53969), sin_addr=inet_addr("10.244.0.74")}, msg_namelen=112 => 16, msg_iov=[{iov_base="\305\377\1\0\0\1\0\0\0\0\0\1\7googlel\3com\0\0\1\0\1\0\0)"..., iov_len=512}], msg_iovlen=1, msg_control=[{cmsg_len=28, cmsg_level=SOL_IP, cmsg_type=IP_PKTINFO, cmsg_data={ipi_ifindex=9, ipi_spec_dst=inet_addr("127.0.0.11"), ipi_addr=inet_addr("127.0.0.11")}}], msg_controllen=32, msg_flags=0}, 0) = 40
1154  futex(0xc0001a8948, FUTEX_WAKE_PRIVATE, 1) = 1
1154  recvmsg(69<UDP:[691492]>,  <unfinished ...>
1154  <... recvmsg resumed>{msg_namelen=112}, 0) = -1 EAGAIN (Resource temporarily unavailable)
1154  setns(14<net:[4026531840]>, CLONE_NEWNET <unfinished ...>
1154  <... setns resumed>)              = 0
1154  openat(AT_FDCWD</>, "/var/run/docker/netns/acf7be30eece", O_RDONLY|O_CLOEXEC <unfinished ...>
1154  <... openat resumed>)             = 26</run/docker/netns/acf7be30eece>
1154  setns(26</run/docker/netns/acf7be30eece>, CLONE_NEWNET) = 0
1154  socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 28<UDP:[3313658]>
1154  setsockopt(28<UDP:[3313658]>, SOL_SOCKET, SO_BROADCAST, [1], 4 <unfinished ...>
1154  <... setsockopt resumed>)         = 0
1154  connect(28<UDP:[3313658]>, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, 16 <unfinished ...>
1154  <... connect resumed>)            = 0
1154  epoll_ctl(5<anon_inode:[eventpoll]>, EPOLL_CTL_ADD, 28<UDP:[3313658]>, {events=EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, data={u32=3036353360, u64=139898711118672}} <unfinished ...>
1154  <... epoll_ctl resumed>)          = 0
1154  getsockname(28<UDP:[3313658]>,  <unfinished ...>
1154  <... getsockname resumed>{sa_family=AF_INET, sin_port=htons(39684), sin_addr=inet_addr("172.12.1.4")}, [112 => 16]) = 0
1154  getpeername(28<UDP:[3313658]>,  <unfinished ...>
1154  <... getpeername resumed>{sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, [112 => 16]) = 0
1154  setns(14<net:[4026531840]>, CLONE_NEWNET) = 0
1154  close(26</run/docker/netns/acf7be30eece> <unfinished ...>
1154  <... close resumed>)              = 0
1154  write(28<UDP:[3313658]>, "\305\377\1\0\0\1\0\0\0\0\0\1\7googlel\3com\0\0\1\0\1\0\0)"..., 40 <unfinished ...>
1154  <... write resumed>)
```

It looks that the thread have entered the netns of the `kind-control-plane`, and then did the DNS resolution.

The missing bit in the puzzle is how the following got redirected to the `dockerd` socket w/o leaving the netns (`dockerd` runs in the host netns, while `kind-control-plane` in its own):
```
lxc61ce9e2c0db2 In  IP 10.244.0.74.40127 > 172.12.1.1.53: 57830+ [1au] A? googlel.com. (52)
```

**Comment by brb** [comment]
There we go, the last missing puzzle with the help of pwru:
```
0xffff92c424982a00      2        [coredns]  nf_csum_update	[nf_nat] sec-arg=0 netns=4026533199 mark=0x10970f00 ifindex=7 proto=8 mtu=1500 len=68 10.244.0.195:60074->172.12.1.1:53(udp)
0xffff92c424982a00      2        [coredns] inet_proto_csum_replace4 sec-arg=18446623969858038272 netns=4026533199 mark=0x10970f00 ifindex=7 proto=8 mtu=1500 len=68 10.244.0.195:60074->172.12.1.1:53(udp)
0xffff92c424982a00      2        [coredns] inet_proto_csum_replace4 sec-arg=18446623969858038272 netns=4026533199 mark=0x10970f00 ifindex=7 proto=8 mtu=1500 len=68 10.244.0.195:60074->172.12.1.1:53(udp)
0xffff92c424982a00      2        [coredns]       udp_v4_early_demux sec-arg=18446623969858038272 netns=4026533199 mark=0x10970f00 ifindex=7 proto=8 mtu=1500 len=68 10.244.0.195:60074->127.0.0.11:38359(udp)
```

Which happens because of the following rule:
```
[222:12858] -A DOCKER_OUTPUT -d 172.12.1.1/32 -p udp -j DNAT --to-destination 127.0.0.11:38359
```

**Comment by brb** [comment]
TL;DR because the BPF host makes the packet to bypass that iptables rule, the packet is actually sent to the host netns. As there is no UDP socket listening on `:53`, the packet gets dropped.

**Comment by gandro** [comment]
A note for people stumbling upon this issue like I have: 

This has only been "fixed" when using our `kind.sh` script. If one uses plain kind or the `helm/kind-action` GitHub action (such as we do in some, but not all, GitHub workflows), then one still has to disable BPF host routing for the connectivity test to pass.
